plugins/synapticsmst/synapticsmst-device.c

```diff
@@ -586,139 +1094,242 @@ synapticsmst_device_write_firmware (SynapticsMSTDevice *device,
 				     "configuration checksum error");
 		return FALSE;
 	}
-
+	/* Firmware Size */
 	checksum = 0;
 	offset = 0x400;
-	code_size = (*(payload_data + offset) << 8) + *(payload_data + offset + 1);
-	if (code_size >= 0xFFFF) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "invalid firmware size");
-		return FALSE;
+	if(isChipPanamera == TRUE){
+		code_size = (*(payload_data + offset) << 24);
+		code_size += (*(payload_data + offset + 1)<<16);
+		code_size += (*(payload_data + offset + 2)<<8);
+		code_size += (*(payload_data + offset + 3));
+		
+		if (code_size >= 0x02FFFF) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"invalid firmware size");
+			return FALSE;
+		}
 	}
-	for (guint32 i = 0; i < (code_size + 17); i++)
-		checksum += *(payload_data + offset + i);
-
-	if (checksum & 0xFF) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "firmware checksum error");
-		return FALSE;
+	else{
+		code_size = (*(payload_data + offset) << 8) + *(payload_data + offset + 1);
+		if (code_size >= 0xFFFF) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"invalid firmware size");
+			return FALSE;
+		}
 	}
+	/* Firmware Checksum */
+	if(isChipPanamera == TRUE){
+		guint32 i =0;
+		for( i = 0; i < code_size + 1; i++ )
+		{
+			checksum += *(payload_data + offset + 0x10 + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x408) << 8) | (*(payload_data + 0x409)))) {
+			g_set_error_literal (error,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+		checksum = 0;
+		offset = EEPROM_ESM_OFFSET;
+		for (i = 0; i < (payload_len - EEPROM_ESM_OFFSET); i++)
+		{
+			checksum += *(payload_data + offset + i);
+		}
+		if ((checksum & 0xFFFF)!=((*(payload_data + 0x40A) << 8) | (*(payload_data + 0x40B)))) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+							"ESM firmware checksum error");
+			return FALSE;
+		}		
+	}
+	else{	
+		for (guint32 i = 0; i < (code_size + 17); i++)
+			checksum += *(payload_data + offset + i);
 
+		if (checksum & 0xFF) {
+			g_set_error_literal (error,	G_IO_ERROR,	G_IO_ERROR_INVALID_DATA,
+						"firmware checksum error");
+			return FALSE;
+		}
+	}
 	/* TODO: May need a way to override this to cover field
 	 * issues of invalid firmware flashed*/
 	/* check firmware and board ID again */
 	tmp = (*(payload_data + ADDR_CUSTOMER_ID) << 8) + *(payload_data + ADDR_BOARD_ID);
-	if (tmp != synapticsmst_device_get_board_id (device)) {
-		g_set_error_literal (error,
-				     G_IO_ERROR,
-				     G_IO_ERROR_INVALID_DATA,
-				     "board ID mismatch");
-		return FALSE;
+	if(isChipPanamera == TRUE){
+		g_debug("bypass check CusID %x %x",tmp, synapticsmst_device_get_board_id (device));	
```

Comment:
Is this right?  I would think you'll still want to do customer ID check on Panamera.

Reply:
Also, `if(isChipPanamera == TRUE){` -> `if (is_chip_panamera) {` -- I know I'm fairly anal about style, but if it's one codebase it really has to be consistent and easy to read.

Reply:
This is because the EVB demo FW didn't fill this part, this should modify and test with customer's FW. Not done yet.

Reply:
It would be better to just test if this is EVB FW and skip the check only in that instance.